### PR TITLE
11.0 valuation at date wizard

### PR DESCRIPTION
**Pull request by sle-odoo**

### CODE REVIEW

**Comment by sle-odoo** [comment]
@hbto @wtaferner heads-up

File: addons/stock_account/models/product.py [hunk 1]
```diff
@@ -176,13 +182,53 @@ def _sum_remaining_values(self):
         return sum(moves.mapped('remaining_value'))
 
     @api.multi
-    @api.depends('stock_move_ids.product_qty', 'stock_move_ids.state', 'stock_move_ids.remaining_value', 'product_tmpl_id.cost_method', 'product_tmpl_id.standard_price')
+    @api.depends('stock_move_ids.product_qty', 'stock_move_ids.state', 'stock_move_ids.remaining_value', 'product_tmpl_id.cost_method', 'product_tmpl_id.standard_price', 'product_tmpl_id.property_valuation', 'product_tmpl_id.categ_id.property_valuation')
     def _compute_stock_value(self):
+        StockMove = self.env['stock.move']
+        to_date = self.env.context.get('to_date')
+
+        fifo_automated_values = {}
```
Comment (sle-odoo) [inline addons/stock_account/models/product.py]: probably add a _check_access_rules here

File: addons/stock_account/views/stock_account_views.xml [hunk 2]
```diff
@@ -49,6 +49,82 @@
...
+            <field name="res_model">stock.quantity.history</field>
+            <field name="view_type">form</field>
+            <field name="view_mode">form</field>
+            <field name="view_id" ref="stock_account.view_stock_quantity_history"/>
+            <field name="target">new</field>
+            <field name="context">{'default_compute_at_date': 0, 'valuation': True}</field>
+        </record>
+        <menuitem id="menu_valuation" name="Inventory Valuation" parent="stock.menu_warehouse_report" sequence="110" action="action_stock_inventory_valuation"/>
+
+        <!-- AML specialized tree view -->
```
Comment (sle-odoo) [inline addons/stock_account/views/stock_account_views.xml]: probably add a higher priority

File: addons/stock_account/models/product.py [hunk 3]
```diff
@@ -176,13 +182,75 @@ def _sum_remaining_values(self):
...
+        if to_date:
+            query = """SELECT aml.product_id, aml.account_id, sum(aml.debit) - sum(aml.credit), sum(quantity), array_agg(aml.id)
+                         FROM account_move_line AS aml
+                   INNER JOIN account_move AS am ON (aml.move_id = am.id)
+                        WHERE am.stock_move_id IS NOT NULL AND am.date <= %s
+                     GROUP BY aml.product_id, aml.account_id"""
+            params = (to_date,)
```
Comment (sle-odoo) [inline addons/stock_account/models/product.py]: in the WHERE clause filter on aml account_id current and previous valuation's account
tricky in sql (join with categ_id and ir.property and play around the res_value column) so a previous sql/orm query will do the trick

without this change if the user updates its valuation account the value aren't correct

**Comment by hbto** [comment]
Thanks @sle-odoo I will be checking it.

Regards

**Comment by wtaferner** [comment]
Thank you for the info @sle-odoo, much appreciated - not sure what you are trying to fix, but reading the commit message tells me that it is a rather good thing for merging ```stock.move``` in case the standard costs have changed and you get stuck or the move is not merged just b/c of the different costs on the stock move (this one I experienced quite some time but did ignore most of the times).

From an accounting perspective of course it is also better to have the correct costs with the stock moves, but honestly I was never using or recommending to use for valuation under standard costs as it is simpler and more reliable to calculate based on the current stock level and to do the math by yourself and manually post the accounting entries ;-)

So, probably I am not of a big help here, but let me know when this is ready to merge that I will test it on my instances in order to see how such a massive change works with "real data" and how smooth and consistent the upgrade works.

File: addons/stock_account/wizard/stock_quantity_history.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+class StockQuantityHistory(models.TransientModel):
+    _inherit = 'stock.quantity.history'
+
+    def open_table(self):
+        if not self.env.context.get('valuation'):
+            return super(StockQuantityHistory, self).open_table()
+
+        self.env['stock.move']._run_fifo_vacuum()
+
+        if self.compute_at_date:
```
[the remark this answers is not in the document]
Comment (sle-odoo) [reply addons/stock_account/wizard/stock_quantity_history.py]: it's done in the latter commits

File: addons/stock_landed_costs/models/stock_landed_cost.py [hunk 5]
```diff
@@ -103,6 +103,7 @@ def button_validate(self):
                 new_landed_cost_value = line.move_id.landed_cost_value + line.additional_landed_cost
                 line.move_id.write({
                     'landed_cost_value': new_landed_cost_value,
+                    'value': line.move_id.value + cost_to_add,
```
Comment (jco-odoo) [inline addons/stock_landed_costs/models/stock_landed_cost.py]: In the migration scripts, I supposed the value was without landed costs, so I subtracted it again.  Need to be clear about this.

**Pull request merged**